token-swap/program/src/curve/stable.rs

```diff
@@ -59,20 +86,18 @@ fn calculate_step(
 /// Compute stable swap invariant (D)
 /// Equation:
 /// A * sum(x_i) * n**n + D = A * D * n**n + D**(n+1) / (n**n * prod(x_i))
-fn compute_d(amp: u128, amount_a: u128, amount_b: u128) -> Option<u128> {
-    // XXX: Curve uses u256
-    let n_coins: u128 = 2; // n
-    let amount_a_times_coins = amount_a.checked_mul(n_coins)?;
-    let amount_b_times_coins = amount_b.checked_mul(n_coins)?;
+fn compute_d(leverage: u64, amount_a: u128, amount_b: u128) -> Option<u128> {
+    let amount_a_times_coins = checked_u8_mul(&amount_a.into(), N_COINS)?;
+    let amount_b_times_coins = checked_u8_mul(&amount_b.into(), N_COINS)?;
     let sum_x = amount_a.checked_add(amount_b)?; // sum(x_i), a.k.a S
     if sum_x == 0 {
         Some(0)
     } else {
-        let mut d_previous: u128;
-        let mut d = sum_x;
-        let leverage = amp.checked_mul(n_coins)?; // A * n
-                                                  // Newton's method to approximate D
-        for _ in 0..128 {
+        let mut d_previous: U256;
+        let mut d: U256 = sum_x.into();
+
+        // Newton's method to approximate D
+        for _ in 0..32 {
```

Comment:
Did you do some instruction number checking for this, or is it based on our previous discussions?  I'd be curious to know how `U256` performs!

Reply:
Here is the method I used: added all those range tests, got an overflow error and moved parts of the code to U256.
Once the overflow was resolved I reduced the number of iterations from 128, to 64, then to 32, then 16. On 16 tests broke so I kept 32.

Reply:
awesome, that sounds reasonable. we'll see how this does against the fuzzing!